Dockerfile.build

```diff
@@ -0,0 +1,18 @@
+FROM golang:1.12 AS builder
+
+ENV CGO_ENABLED=0
+
+WORKDIR /build
+
+COPY go.mod /build/go.mod
+RUN go mod download
+
+COPY cmd /build/cmd
```

Comment:
It'll look a little silly but have used

````
COPY . .
```

in cases like this to stage the whole project source after doing the dependencies download stage.

I'm not sure if I like the `pkg` convention either since that forces an extra `pkg` level into the module path of `github.com/itzg/mc-router/pkg/server` as an example in this case.

Reply:
My issue with `COPY .` is that it invalidates the cache if anything changes in that directory, even if it has nothing to do with the actual build. As a dumb example, if I have my Dockerfile:
```
FROM golang:1.12 AS builder

COPY . /build
RUN go build -o mc-router /build/cmd/mc-router/main.go

FROM scratch

COPY --from=builder /build/mc-router /bin/mc-router

ENTRYPOINT ["/bin/mc-router"]
```
and I want to add an environment variable to my container, like so:
```
FROM golang:1.12 AS builder

COPY . /build
RUN go build -o mc-router /build/cmd/mc-router/main.go

FROM scratch

COPY --from=builder /build/mc-router /bin/mc-router
ENV EXAMPLE=true

ENTRYPOINT ["/bin/mc-router"]
```

Rather than properly caching and starting at the second to last step in the second container, it will go all the way back to my build container and start over from the `COPY .` task.

All that being said it's a fairly minor issue especially for a project of this size, just wanted to explain my reasoning before I reverted it. If you still prefer the top-level style rather than the `pkg` style I can move them back.

Reply:
First of all, thank you for entertaining my questioning. Like you said, for a project this size it's not really a big deal either way.

however, I'm still leaning towards the top-level style and the original `Dockerfile` because I have been refining and using the same CI work flow with goreleaser and CircleCI over several projects. 

I'd still like the sound of an adjacent, development-time `Dockerfile.build` or similar and would clarify my suggestion to have that include the section:

```
WORKDIR /build
COPY go.mod go.sum .
RUN go mod download
COPY . .
RUN go build -o mc-router cmd/mc-router/main.go
```

It doesn't solve all the cache invalidation scenarios, but will at least cache the module downloading between iterations.

Reply:
Sorry, I just noticed you did put this in `Dockerfile.build`, but the two step COPY still applies :)